### PR TITLE
Hybrid filters

### PR DESCRIPTION
**Pull request by donomii**
### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: usecases/traverser/hybrid/searcher.go [hunk 1]
```diff
@@ -107,37 +107,34 @@ func (s *Searcher) Search(ctx context.Context) (Results, error) {
...
-			if err != nil {
-				return nil, err
-			}
+			if err == nil {
 
-			found = append(found, res)
-			weights = append(weights, 1-alpha)
+				found = append(found, res)
+				weights = append(weights, 1-alpha)
+			}
```
Comment (parkerduckworth) [inline usecases/traverser/hybrid/searcher.go:114]: Is this error (and the 2 below) just being ignored? or is it propagating somewhere?

File: adapters/repos/db/ranked_fusion_test.go [hunk 2]
```diff
@@ -456,6 +457,185 @@ func TestRFJourney(t *testing.T) {
...
+				Query:  "elephant",
+				Vector: elephantVector(),
+				Limit:  100,
+				Alpha:  0.5,
+			},
+			Pagination: &filters.Pagination{
+				Offset: 0,
+				Limit:  -1,
+			},
+			// Filters: filter,
```
Comment (parkerduckworth) [inline adapters/repos/db/ranked_fusion_test.go:615]: Should this be removed?

File: adapters/repos/db/inverted/bm25_searcher.go [hunk 3]
```diff
@@ -301,20 +303,37 @@ func (b *BM25Searcher) getTopKHeap(limit int, results terms, averagePropLength f
 	}
 }
 
-func (b *BM25Searcher) createTerm(N float64, query string, propertyNames []string, propertyBoosts map[string]float32, duplicateTextBoost int) (term, map[uint64]int, error) {
+func (b *BM25Searcher) createTerm(N float64, filterDocIds helpers.AllowList, query string, propertyNames []string, propertyBoosts map[string]float32, duplicateTextBoost int) (term, map[uint64]int, error) {
 	var docMapPairs []docPointerWithScore = nil
 	var docMapPairsIndices map[uint64]int = nil
 	termResult := term{queryTerm: query}
+	uniqeDocIDs := sroar.NewBitmap() // to build the global n if there is a filter
```
Comment (parkerduckworth) [inline adapters/repos/db/inverted/bm25_searcher.go:310]: a typo maybe: uniqeDocIDs -> uniqueDocIDs

File: adapters/repos/db/inverted/bm25_searcher.go [hunk 4]
```diff
@@ -21,12 +21,14 @@ import (
...
 	"github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/priorityqueue"
 
 	"github.com/weaviate/weaviate/entities/models"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+
 	"github.com/weaviate/weaviate/adapters/repos/db/helpers"
 	"github.com/weaviate/weaviate/adapters/repos/db/lsmkv"
 	"github.com/weaviate/weaviate/adapters/repos/db/propertyspecific"
```
Comment (parkerduckworth) [inline adapters/repos/db/inverted/bm25_searcher.go:34]: sort please

File: adapters/repos/db/shard_read.go [hunk 5]
```diff
@@ -21,6 +21,8 @@ import (
 	"github.com/google/uuid"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+
+	//"github.com/weaviate/sroar"
```
Comment (parkerduckworth) [inline adapters/repos/db/shard_read.go:25]: Should this be removed?

File: usecases/traverser/hybrid/searcher.go (hunk 1 above)
Comment (donomii) [reply usecases/traverser/hybrid/searcher.go:114]: They are being discarded

File: adapters/repos/db/ranked_fusion_test.go (hunk 2 above)
Comment (donomii) [reply adapters/repos/db/ranked_fusion_test.go:615]: Why would we remove it?

File: adapters/repos/db/inverted/bm25_searcher.go (hunk 3 above)
Comment (donomii) [inline adapters/repos/db/inverted/bm25_searcher.go:310]:
```suggestion
	uniqueDocIDs := sroar.NewBitmap() // to build the global n if there is a filter
```

File: usecases/traverser/hybrid/searcher.go (hunk 1 above)
Comment (donomii) [reply usecases/traverser/hybrid/searcher.go:114]: Error propogation will require modifying quite. few functions, we'll do it for the next release

File: adapters/repos/db/ranked_fusion_test.go [hunk 6]
```diff
@@ -456,6 +457,184 @@ func TestRFJourney(t *testing.T) {
...
+
+		metrics := &fakeMetrics{}
+		log, _ := test.NewNullLogger()
+		explorer := traverser.NewExplorer(repo, log, prov, metrics)
+		hybridResults, err := explorer.Hybrid(context.TODO(), params)
+		require.Nil(t, err)
+		require.Equal(t, 0, len(hybridResults))
+	})
+
+	t.Run("Hybrid with filter", func(t *testing.T) {
```
Comment (dirkkul) [inline adapters/repos/db/ranked_fusion_test.go:602]: could you adapt the names to what the tests are doing? The one above is named "hybrid", but has filters and this one is named "Hybrid with filter" but has no filters

**Review by dirkkul — APPROVED** [review]

**Pull request merged**